src/Integrators.jl

```diff
@@ -175,6 +246,114 @@ function evolve!(integrator::SphericalMidpoint, Δt::Float64)
 end
 
 
+""" SU(N) integrators and helper functions.
+"""
+
+# LinearAlgebra's `normalize!` doesn't normalize to 1 on complex vectors
+function normalize!(Z::Array{CVec{N}, 4}) where N
+    for i ∈ eachindex(Z)
+        Z[i] = Z[i] / √(real(Z[i]' * Z[i]))
+    end
+end
+
+
+@inline function _proj(a::T, Z::T)  where T <: CVec
+    (a - ((Z' * a) * Z))  
+end
+
+
+function _apply_ℌ!(rhs, sys, B, Z)
+    aniso = sys.hamiltonian.sun_aniso
+    latsize = size(B)[2:end]
+    for (site, Λ) in zip(aniso.sites, aniso.Λs)
+        for cell ∈ CartesianIndices(latsize)
+             rhs[site, cell] = (-B[site, cell] ⋅ sys.S + Λ) * Z[site, cell] # ∇E = -B
+        end
+    end
+end
+
+
+function _rhs_langevin!(ΔZ, Z, integrator, Δt)
+    (; kT, α, sys, _ℌZ, _ξ, _B) = integrator
+    (; _dipoles) = sys
+
+    set_expected_spins!(_dipoles, Z) # Not needed for Heun as currently written. Keep for safety, consistency? 
+    field!(_B, _dipoles, Z, sys.hamiltonian)
+    _apply_ℌ!(_ℌZ, sys, _B, Z)
+
+    for i ∈ eachindex(Z)
+        ΔZ′ = -im*√(2*Δt*kT*α)*_ξ[i] - Δt*(im+α)*_ℌZ[i]    
+        ΔZ[i] = _proj(ΔZ′, Z[i])
+    end 
+end
+
+
+function _rhs_ll!(ΔZ, Z, integrator, Δt)
+    (; sys, _ℌZ, _B) = integrator
+    (; _dipoles) = sys
+
+    set_expected_spins!(_dipoles, Z) # temporarily de-synchs _dipoles and _coherents
+    field!(_B, _dipoles, Z, sys.hamiltonian)
+    _apply_ℌ!(_ℌZ, sys, _B, Z)
+
+    for i ∈ eachindex(Z)
+        ΔZ[i] = - Δt*im*_ℌZ[i]    
+    end 
+end
```

Comment:
Have a single `rhs` function and dispatch on type?

Reply:
I like it how you wrote it here! By the time we're in `evolve!` we already know for certain what type of integration we're doing, so I don't see a significant benefit to also dispatching on this.